packages/devtools/lib/src/timeline/flame_chart_canvas.dart

```diff
@@ -0,0 +1,624 @@
+  final int index;
+}
+
+// TODO Maybe pass an on selected listener to this so we don't have to pass the data
+// object around
+class FlameChartNode {
+  FlameChartNode(
+    this.rect,
+    this.backgroundColor,
+    this.textColor,
+    this.selectedTextColor,
+    this.stackFrame,
+  )   : startingLeft = rect.left,
+        startingWidth = rect.width;
+
+  static const horizontalPadding = 4.0;
+  static const borderRadius = 2.0;
+  static const selectedBorderWidth = 1.0;
+  static const selectedBorderColor = Color(0x5A1B1F23);
+  static const maxDisplayTextLength = 20;
+
+  Rect rect;
+
+  /// Left value for the flame chart item at zoom level 1.
+  final num startingLeft;
+
+  /// Width value for the flame chart item at zoom level 1;
+  final num startingWidth;
+
+  final Color backgroundColor;
+
+  final Color textColor;
+
+  final Color selectedTextColor;
+
+  final CpuStackFrame stackFrame;
+
+  String get text => stackFrame.name;
+
+  String get tooltip => stackFrame.toString();
+
+  bool selected = false;
+
+  void paint(CanvasRenderingContext2D canvas) {
+    // Fill a rectangle with a border radius.
+    canvas
+      ..fillStyle =
+          colorToCss(selected ? selectedFlameChartNodeColor : backgroundColor)
+      ..beginPath()
+      ..moveTo(rect.left + borderRadius, rect.top)
+      ..lineTo(rect.right - borderRadius, rect.top)
+      ..quadraticCurveTo(
+        rect.right,
+        rect.top,
+        rect.right,
+        rect.top + borderRadius,
+      )
+      ..lineTo(rect.right, rect.bottom - borderRadius)
+      ..quadraticCurveTo(
+        rect.right,
+        rect.bottom,
+        rect.right - borderRadius,
+        rect.bottom,
+      )
+      ..lineTo(rect.left + borderRadius, rect.bottom)
+      ..quadraticCurveTo(
+        rect.left,
+        rect.bottom,
+        rect.left,
+        rect.bottom - borderRadius,
+      )
+      ..lineTo(rect.left, rect.top + borderRadius)
+      ..quadraticCurveTo(
+        rect.left,
+        rect.top,
+        rect.left + borderRadius,
+        rect.top,
+      )
+      ..closePath()
+      ..fill();
+
+    if (selected) {
+      canvas
+        ..strokeStyle = colorToCss(selectedBorderColor)
+        ..lineWidth = selectedBorderWidth
+        ..stroke();
+    }
+
+    canvas
+      ..fillStyle = colorToCss(textColor)
+      ..font = fontStyleToCss(TextStyle(fontSize: fontSize));
+
+    // TODO(kenzie): polish this. Sometimes we trim excessively. We should do
+    // something smarter here to be more exact. 'm' is arbitrary - it was
+    // selected after trial and error with different letter measurements.
+    final singleLetterWidth = canvas.measureText('m').width;
+    final maxLetters = math.min(
+      (rect.width - horizontalPadding * 2) ~/ singleLetterWidth,
+      selected ? text.length : math.min(text.length, maxDisplayTextLength),
+    );
+
+    String displayText = text.substring(0, maxLetters);
+    if (maxLetters == maxDisplayTextLength &&
```

Comment:
add a TODO to measure the text and truncate based on the actual width of the available area.